pubsub/src/manager.rs

```diff
@@ -0,0 +1,88 @@
+//! Provides an executor for subscription Futures.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::sync::Arc;
+
+use crate::core::futures::sync::oneshot;
+use crate::core::futures::{future, Future};
+use crate::{
+	typed::{Sink, Subscriber},
+	SubscriptionId,
+};
+use log::{error, warn};
+use parking_lot::Mutex;
+
+/// Alias for an implementation of `futures::future::Executor`.
+pub type TaskExecutor = Arc<dyn future::Executor<Box<dyn Future<Item = (), Error = ()> + Send>> + Send + Sync>;
+
+/// Trait used to provide unique subscription ids.
+pub trait IdProvider {
+	/// A unique ID used to identify a subscription.
+	type Id: Copy + Clone + Default + Eq + Hash + Into<SubscriptionId> + From<SubscriptionId>;
+
+	/// Returns next id for the subscription.
+	fn next_id(&self) -> Self::Id;
+}
+
+/// Subscriptions manager.
+///
+/// Takes care of assigning unique subscription ids and
+/// driving the sinks into completion.
+#[derive(Clone)]
+pub struct SubscriptionManager<I: Default + IdProvider> {
```

Comment:
```suggestion
pub struct SubscriptionManager<I: IdProvider = RandomizedStringProvider> {
```

We should implement two `IdProviders`:
1. `NumericIdProvider`
2. `RandomizedStringProvider`

The first one would simply use `AtomicU64` and would return increasing numeric values, the second one - which should also be a default to use if `SubscriptionManager` does not include generic parameter - should return a randomized strings (guids).
The reason for using randomized strings is to make sure that subscription ids are not easy to guess, so that multiple clients can't cancel each-others subscriptions easily. The first one will be mostly usable for tests or cases where the aforementioned issue is not a problem.